File: PaMVPDemo/lib-core/src/main/java/com/pa/core/config/NetworkConfig.java
package com.pa.core.config;

import okhttp3.Interceptor;

public class NetworkConfig {

    public final String baseURL;
    public final long readTime;
    public final long writeTime;
    public final long connectTime;
    public final Interceptor[] interceptors;

    private NetworkConfig(Builder builder){
        this.baseURL = builder.baseURL;
        this.readTime = builder.readTime;
        this.writeTime = builder.writeTime;
        this.connectTime = builder.connectTime;
        this.interceptors = builder.interceptors;
    }

    public static class Builder{
        String baseURL;
        long readTime = 30;
        long writeTime = 30;
        long connectTime = 30;
        Interceptor[] interceptors = null;


        public Builder setBaseURL(String baseURL){
            this.baseURL = baseURL;
            return this;
        }

        public Builder setReadTime(long readTime){
            this.readTime = readTime;
            return this;
        }

        public Builder setWriteTime(long writeTime){
            this.writeTime = writeTime;
            return this;
        }

        public Builder setConnectTime(long connectTime){
            this.connectTime = connectTime;
            return this;
        }

        public Builder setInterceptor(Interceptor... interceptors){
            this.interceptors = interceptors;
            return this;
        }

        public NetworkConfig build(){
            return new NetworkConfig(this);
        }
    }
}

File: PaMVPKotlinDemo/lib-core/src/main/java/com/pa/core/config/NetworkConfig.kt
package com.pa.core.config

import okhttp3.Interceptor

class NetworkConfig private constructor(builder: Builder) {

    val baseURL: String?
    val readTime: Long
    val writeTime: Long
    val connectTime: Long
    val interceptors: Array<Interceptor>?

    //新的关键字init用来处理类的初始化问题，init模块中的内容可以直接使用构造函数的参数
    init {
        this.baseURL = builder.baseURL
        this.readTime = builder.readTime
        this.writeTime = builder.writeTime
        this.connectTime = builder.connectTime
        this.interceptors = builder.interceptors
    }

    class Builder {

        //lateinit只能在不可null的对象上使用，比须为var，不能为primitives（Int、Float之类）
        var baseURL: String? = null
        var readTime: Long = 30
        var writeTime: Long = 30
        var connectTime: Long = 30
        var interceptors: Array<Interceptor>? = null


        fun setBaseURL(baseURL: String): Builder {
            this.baseURL = baseURL
            return this
        }

        fun setReadTime(readTime: Long): Builder {
            this.readTime = readTime
            return this
        }

        fun setWriteTime(writeTime: Long): Builder {
            this.writeTime = writeTime
            return this
        }

        fun setConnectTime(connectTime: Long): Builder {
            this.connectTime = connectTime
            return this
        }

        //vararg：可变参数
        fun setInterceptor(interceptors: Array<Interceptor>?): Builder {
            this.interceptors = interceptors
            return this
        }

        fun build(): NetworkConfig {
            return NetworkConfig(this)
        }
    }
}
